### PR TITLE
[SPARK-19751][SQL] Throw an exception if bean class has one's own class in fields

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
The current master throws `StackOverflowError` in `createDataFrame`/`createDataset` if bean has one's own class in fields;
```
public class SelfClassInFieldBean implements Serializable {
  private SelfClassInFieldBean child;
  ...
}
```
This pr added code to throw `UnsupportedOperationException` in that case as soon as possible.

## How was this patch tested?
Added tests in `JavaDataFrameSuite` and `JavaDatasetSuite`.

### CODE REVIEW

**Comment by srowen** [comment]
This at least improves the error, so I think it's OK to merge at least. I am not sure whether this can work or whether it must be an error. CC @cloud-fan maybe

**Comment by cloud-fan** [comment]
We don't support circle reference in class hierarchy, so this case should fail.

But this PR only checks if a field is same-type of the class, not for all circles. e.g.
```
class A {
  B b;
}

class B {
  A a;
}
```
will still throw StackOverflowError

**Comment by maropu** [comment]
okay, I'll update soon.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 1]
```diff
@@ -69,7 +69,8 @@ object JavaTypeInference {
    * @param typeToken Java type
    * @return (SQL data type, nullable)
    */
-  private def inferDataType(typeToken: TypeToken[_]): (DataType, Boolean) = {
+  private def inferDataType(typeToken: TypeToken[_], seenTypeSet: Set[Class[_]] = Set.empty)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: does scala case class have the same problem?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: You mean this case?
```

scala> :paste
case class classA(i: Int, cls: classB)
case class classB(cls: classA)

scala> Seq(classA(0, null)).toDS()
java.lang.StackOverflowError
  at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1494)
  at scala.reflect.runtime.JavaMirrors$JavaMirror$$anon$1.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$info(JavaMirrors.scala:66)
  at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anonfun$info$1.apply(SynchronizedSymbols.scala:127)
  at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anonfun$info$1.apply(SynchronizedSymbols.scala:127)
  at scala.reflect.runtime.Gil$class.gilSynchronized(Gil.scala:19)
  at scala.reflect.runtime.JavaUniverse.gilSynchronized(JavaUniverse.scala:16)
  at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$class.gilSynchronizedIfNotThreadsafe(SynchronizedSymbols.scala:123)
  at scala.reflect.runtime.JavaMirrors$JavaMirror$$anon$1.gilSynchronizedIfNotThreadsafe(JavaMirrors.scala:66)
  at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$class.info(SynchronizedSymbols.scala:127)
  at scala.reflect.runtime.JavaMirrors$JavaMirror$$anon$1.info(JavaMirrors.scala:66)
  at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:48)
  at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:45)
  at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:45)
  at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:45)
  at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:45)
```

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala (hunk 1 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: I filed JIRA for this case: https://issues.apache.org/jira/browse/SPARK-19896

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 2]
```diff
@@ -123,7 +124,12 @@ object JavaTypeInference {
         val properties = getJavaBeanReadableProperties(other)
         val fields = properties.map { property =>
           val returnType = typeToken.method(property.getReadMethod).getReturnType
-          val (dataType, nullable) = inferDataType(returnType)
+          if (seenTypeSet.contains(returnType.getRawType)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: what about `class A(a: Array[A])`?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: I'll also check other complex types.

**Comment by maropu** [comment]
@cloud-fan Could you check again? Thanks!

**Comment by maropu** [comment]
@cloud-fan ping

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 3]
```diff
@@ -123,13 +124,41 @@ object JavaTypeInference {
         val properties = getJavaBeanReadableProperties(other)
         val fields = properties.map { property =>
           val returnType = typeToken.method(property.getReadMethod).getReturnType
-          val (dataType, nullable) = inferDataType(returnType)
+          val types = extractRawTypes(returnType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: This looks fragile, how about something like
```
case other =>
  if (seenTypeSet.contains(other)) throw xxx
  ...
  ....  inferDataType(returnType, seenTypeSet ++ other)

```

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: okay, I'll update based you suggestion.

**Comment by cloud-fan** [comment]
LGTM, pending test

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**